pkg/controller/logstorage/initializer/conditions_controller.go

```diff
@@ -90,18 +91,272 @@ func (r *LogStorageConditions) Reconcile(ctx context.Context, request reconcile.
+// to compare with the current logStorage conditions.
+func transformIntoLogStorageConditions(desiredConditions map[operatorv1.StatusConditionType]operatorv1.TigeraStatusCondition) map[string]metav1.Condition {
+
+	lsConditions := make(map[string]metav1.Condition)
+
+	// if Degraded or Progressing is true, then Available should be marked as false even
+	// if tigerastatus have both Available and Degraded/Progressing set to True.
+	isDegOrProg := false
+
+	// Check for Degraded
+	if degradedCondition, ok := desiredConditions[operatorv1.ComponentDegraded]; ok {
+		observedGeneration := degradedCondition.ObservedGeneration
+		if degradedCondition.Status == operatorv1.ConditionTrue {
+			isDegOrProg = true
+			lsConditions[string(operatorv1.ComponentDegraded)] = setDegraded(degradedCondition)
+			lsConditions[string(operatorv1.ComponentReady)] = clearAvailable(observedGeneration)
+		} else {
+			lsConditions[string(operatorv1.ComponentDegraded)] = clearDegraded(string(operatorv1.Unknown), "", observedGeneration)
+		}
+	}
+
+	// Check for Progressing
+	if progressingCondition, ok := desiredConditions[operatorv1.ComponentProgressing]; ok {
+		observedGeneration := progressingCondition.ObservedGeneration
+		if progressingCondition.Status == operatorv1.ConditionTrue {
+			isDegOrProg = true
+			lsConditions[string(operatorv1.ComponentProgressing)] = setProgressing(progressingCondition)
+			lsConditions[string(operatorv1.ComponentReady)] = clearAvailable(observedGeneration)
+		} else {
+			reason := string(operatorv1.Unknown)
+			// if degraded and no progressing components set reason as ResourceNotReady
+			if isDegOrProg {
+				reason = string(operatorv1.ResourceNotReady)
+			}
+			lsConditions[string(operatorv1.ComponentProgressing)] = clearProgressing(reason, "", observedGeneration)
+		}
+	}
+
+	// Check for Available
+	if availableCondition, ok := desiredConditions[operatorv1.ComponentAvailable]; ok && !isDegOrProg {
+		// Set Available as true and clear degraded and progressing
+		observedGeneration := availableCondition.ObservedGeneration
+		if availableCondition.Status == operatorv1.ConditionTrue {
+			lsConditions[string(operatorv1.ComponentReady)] = setAvailable(availableCondition)
+			lsConditions[string(operatorv1.ComponentProgressing)] = clearProgressing(string(operatorv1.AllObjectsAvailable), "All objects are available", observedGeneration)
+			lsConditions[string(operatorv1.ComponentDegraded)] = clearDegraded(string(operatorv1.AllObjectsAvailable), "All objects are available", observedGeneration)
+		} else {
+			lsConditions[string(operatorv1.ComponentReady)] = clearAvailable(observedGeneration)
+		}
+	}
+
+	return lsConditions
+}
+
+func updateConditions(currentConditions, desiredConditions map[string]metav1.Condition) []metav1.Condition {
+
+	statusCondition := []metav1.Condition{}
+
+	//Update the current log storage condition only when the aggregate desired status have new changes
+	for _, desired := range desiredConditions {
```

Comment:
One of the reasons I suggested using a `map` type here is that it avoids the need for the inner loop here (and also makes it easier to avoid duplicate entries of the same type)

For example:

```go
for conditionType, desired := range desiredConditions {

  // Get the current condition, if it exists.
  current, ok := currentConditions[conditionType]

  // Update the desired state with the correct transition timestamp. 
  // If the current Status doesn't match the desired,
  // or if there is no current Status, use the current time.
  // Otherwise, use the existing transition timestamp.
  if !ok || current.Status != desired.Status {
    desired.LastTransitionTime = time.Now()
  } else {
    desired.LastTransitionTime = current.LastTransitionTime
  }
  statusConditions = append(statusConditions, desired)
}